src/Components/Server/src/Circuits/RemoteJSDataStream.cs

```diff
@@ -0,0 +1,210 @@
+        private readonly long _streamId;
+        private readonly long _totalLength;
+        private readonly CancellationToken _streamCancellationToken;
+        private readonly Timer _receiveDataTimer;
+        private readonly Stream _pipeReaderStream;
+        private readonly Pipe _pipe;
+        private long _bytesRead;
+        private long _expectedChunkId;
+
+        public static async Task<bool> ReceiveData(RemoteJSRuntime runtime, long streamId, long chunkId, byte[] chunk, string error)
+        {
+            if (!runtime.RemoteJSDataStreamInstances.TryGetValue(streamId, out var instance))
+            {
+                // There is no data stream with the given identifier. It may have already been disposed.
+                // We notify JS that the stream has been cancelled/disposed.
+                return false;
+            }
+
+            return await instance.ReceiveData(chunkId, chunk, error);
+        }
+
+        public static async ValueTask<RemoteJSDataStream> CreateRemoteJSDataStreamAsync(
+            RemoteJSRuntime runtime,
+            IJSStreamReference jsStreamReference,
+            long totalLength,
+            long maxBufferSize,
+            long maximumIncomingBytes,
+            CancellationToken cancellationToken = default)
+        {
+            // Enforce minimum 1 kb SignalR message size as we budget 512 bytes
+            // overhead for the transfer, thus leaving at least 512 bytes for data
+            // transfer per chunk.
+            var chunkSize = maximumIncomingBytes > 1024 ?
+                maximumIncomingBytes - 512 :
+                throw new ArgumentException($"SignalR MaximumIncomingBytes must be at least 1 kb.");
+
+            var streamId = runtime.RemoteJSDataStreamNextInstanceId++;
+            var remoteJSDataStream = new RemoteJSDataStream(runtime, streamId, totalLength, maxBufferSize, cancellationToken);
+            await runtime.InvokeVoidAsync("Blazor._internal.sendJSDataStream", jsStreamReference, streamId, chunkSize);
+            return remoteJSDataStream;
+        }
+
+        private RemoteJSDataStream(
+            RemoteJSRuntime runtime,
+            long streamId,
+            long totalLength,
+            long maxBufferSize,
+            CancellationToken cancellationToken)
+        {
+            _runtime = runtime;
+            _streamId = streamId;
+            _totalLength = totalLength;
+            _streamCancellationToken = cancellationToken;
+
+            // Dispose of the stream if a chunk isn't received within 1 minute.
+            _receiveDataTimer = new Timer(new TimerCallback(async (state) => {
+                var stream = state as RemoteJSDataStream;
+                var timeoutException = new TimeoutException("Did not receive any data in the alloted time.");
+                await stream.CompletePipeAndDisposeStream(timeoutException);
+            }), this, dueTime: ReceiveDataTimeout, period: Timeout.InfiniteTimeSpan);
```

Comment:
There's a better way to do this, we don't need a timer I believe.

You can setup a task with `Task.Delay(1000)` (Use the JS interop timeout from the options) and you need to do this every time we receive data.

Then on ReceiveData "fire and forget a task"

_lastReceivedMessageTime = DateTime.UtcNow; (or something like this);
_ = CancelOnTimeout(_lastReceivedMessageTime);

Reply:
On the Task you check if you received data before the Task.Delay and terminate the circuit (throwing an exception should work).

Reply:
Not sending data fast enough is not a "legitimate" issue.

Reply:
> throwing an exception should work

Throwing an exception from the task isn't sufficient to crush the circuit. However, we can raise an `UnhandledException` like we do within the renderer. Took this approach in 3cd7619020.